### PR TITLE
[TH2-2109] Remove generation xml tags for empty messages.

### PR DESCRIPTION
**Pull request by Ilya-Loginov-Exactpro**
Add ability to create embedded messages collections.

### CODE REVIEW

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt [hunk 1]
```diff
@@ -119,13 +129,31 @@ class XmlMessageStructureVisitor(private val document: Document, private val nod
...
+
+                MessageStructureWriter.WRITER.traverse(
+                    XmlMessageStructureVisitor(
+                        document,
+                        newNode,
+                        messageValue
+                    ), fldStruct
+                )
+
+                if (generated) {
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt]: value of generated is always false

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt [hunk 2]
```diff
@@ -75,30 +78,37 @@ class XmlMessageStructureVisitor(private val document: Document, private val nod
...
+
         if (fldStruct.isEmbedded()) {
             MessageStructureWriter.WRITER.traverse(
                 XmlMessageStructureVisitor(
                     document,
-                    node.appendChild(document.createElement(fldStruct.getXmlTagName())),
+                    newNode,
                     this.message
-                ), fldStruct
+                ) { generated = true }, fldStruct
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt]: Please compare message's field names and fields structure's field names instead

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

**Comment by Ilya-Loginov-Exactpro** [comment]
Added new attribute `XmlHide`. `Embedded` - XML has a nesting level, the parsed message hasn't one. `XmlHide` - the parsed message has a nesting level, XML hasn't one.

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt [hunk 3]
```diff
@@ -519,12 +539,17 @@ open class XmlPipelineCodec : IPipelineCodec {
...
-         * Boolean attribute. If true fields from current message will be set to parent message
+         * Boolean attribute. If true fields from current message will be move to a parent message
          */
         const val EMBEDDED_ATTRIBUTE = "Embedded"
 
         /**
-         * XPath expression for find xml nodes
+         * Boolean attribute. If true xml tag for current message will not created
+         */
+        const val XML_HIDE_ATTRIBUTE = "XmlHide"
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt]: Rename to Virtual

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt [hunk 4]
```diff
@@ -519,12 +539,17 @@ open class XmlPipelineCodec : IPipelineCodec {
         const val XML_ATTRIBUTE_VALUE_ATTRIBUTE = "XmlAttributeValue"
 
         /**
-         * Boolean attribute. If true fields from current message will be set to parent message
+         * Boolean attribute. If true fields from current message will be move to a parent message
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt]:
```suggestion
         * Boolean attribute. If true fields from the current message will be moved to a parent message during decoding and packed to sub-message during encoding.
```

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt [hunk 5]
```diff
@@ -519,12 +539,17 @@ open class XmlPipelineCodec : IPipelineCodec {
...
          */
         const val EMBEDDED_ATTRIBUTE = "Embedded"
 
         /**
-         * XPath expression for find xml nodes
+         * Boolean attribute. If true xml tag for current message will not created
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt]:
```suggestion
         * Boolean attribute. If true fields from the current message will be packed to sub-message during decoding and moved to a parent message during encoding.
```

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

File: src/test/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodecTest.kt [hunk 6]
```diff
@@ -277,6 +278,34 @@ class XmlPipelineCodecTest {
...
+        check("""
+            <TestHideXml>
+                <field0>123</field0>
+                <field1>155</field1>
+            </TestHideXml>
+        """.trimIndent(), message().addFields(
+            "field0", 123,
+            "hide", message().addField("field1", 155)
+        ))
+    }
```
Comment (Nikita-Smirnov-Exactpro) [inline src/test/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodecTest.kt]: please add reverse tests for encode

File: src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt [hunk 7]
```diff
@@ -45,7 +45,13 @@ class XmlMessageStructureVisitor(private val document: Document, private val nod
...
                 }
             } else {
-                node.addNode(fldStruct.getXmlTagName(), document).setText(fieldValue, document)
+                val nodeName = fldStruct.getXmlTagName()
+                val valueNode = node.findNode(nodeName)
+                if (valueNode == null) {
+                    node.addNode(nodeName, document).setText(fieldValue, document)
+                } else if (valueNode.getText() != fieldValue) {
+                    error("Can not encode field with name '$fieldName' with two different values: ['$fieldValue','${valueNode.getText()}']")
+                }
```
Comment (Ilya-Loginov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/codec/xml/XmlMessageStructureVisitor.kt:54]: This check will be executed only when the parent and child virtual message has fields with equals names. But we can prohibit this behavior.

File: src/test/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodecTest.kt [hunk 8]
```diff
@@ -257,67 +247,78 @@ class XmlPipelineCodecTest {
...
-        assertEqualsMessages(message.build(), group.messagesList[0].message)
+        assertEqualsMessages(message.build(), group.messagesList[0].message, true)
     }
 
     private fun assertEqualsMessages(expected: Message, actual: Message, checkMetadata: Boolean = false) {
         if (checkMetadata) {
             assertEquals(expected.metadata.messageType, actual.metadata.messageType, "Not equals message types")
-            assertEquals(expected.metadata.protocol, actual.metadata.protocol, "Not equals protocols")
-            assertEquals(expected.metadata.id, actual.metadata.id, "Not equals ids")
+            //assertEquals(expected.metadata.protocol, actual.metadata.protocol, "Not equals protocols")
```
Comment (Nikita-Smirnov-Exactpro) [inline src/test/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodecTest.kt:320]: Why are these lines commented?

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Pull request merged**